Role allocation

zerver0 opened the pull request:
allow rename requests and checks whether user is in appventure before assigning member role
fixes #3

zerver0 committed:
testing commits

zerver0 committed:
testing commits

zerver0 committed:
add checks on whether the logged in user is from appventure before assigning member role

zerver0 committed:
change identifiers in Member data class to camelcase
replace check if user is in appventure with list.any()
allows exco to use reaction to allow user access to server as guest / alumni or deny access
add 3 fields to config

zerver0 committed:
add excoChannelId field to config
add access request message autodelete when an option was chosen
real name mentioned when access degree chosen

zerver0 committed:
remove override for verify that was added for testing
add notification message if attempted verification user is not in appventure
add nick command to send rename request to exco

zerver0 committed:
fixes nick not detecting the name that was requested

junron commented on src/main/kotlin/app/nush/bot/Config.kt:
```diff
@@ -11,7 +11,11 @@ data class Config(
     val excoRole: String,
-    val dev: Boolean = true
+    val dev: Boolean = true,
+    var userIsAppVentureMember: Boolean = false,
```
This is not necessary

zerver0 replied:
https://github.com/appventure-nush/appventure-bot/pull/7#commitcomment-42151710
you said to put it in the config so :/

junron replied:
That line refers to the [`excoChannelId` line](https://github.com/appventure-nush/appventure-bot/commit/b8afdad94f0070a45fa43bad6f3babe73de9987e#diff-aa9585738278c6cce14b50963e3b3f46R67)

zerver0 replied:
oh

junron commented on src/main/kotlin/app/nush/bot/commands/Nick.kt:
```diff
@@ -0,0 +1,96 @@
+            renameMessage.react("✅")
+            renameMessage.react("❎")
+            reactionAdded { messageReaction ->
+                try {
```
What is this for?

zerver0 replied:
react to the message when it's sent, so exco only has to click, instead of selecting the reaction

junron replied:
I mean the try/catch

zerver0 replied:
well basically if the message is not of the correct type some form of error will be thrown inside the try block then I just ignore the case by putting nothing in catch

junron replied:
I think this is still unnecessary

junron commented on src/main/kotlin/app/nush/bot/commands/Nick.kt:
```diff
@@ -0,0 +1,96 @@
+                try {
+                    if (!guildClient.getMember(messageReaction.userId).user?.isBot!!) {
+                        val newName: String = channelClient.getMessage(messageReaction.messageId).content.subSequence(
+                            channelClient.getMessage(messageReaction.messageId).content.indexOf("\"") + 1,
```
You can use a variable to store `channelClient.getMessage(messageReaction.messageId)` to prevent repeated calls. Try using the intellij shortcut <kbd>Ctrl</kbd> + <kbd>Alt</kbd> + <kbd>v</kbd>
https://www.jetbrains.com/help/idea/extract-variable.html

junron commented on src/main/kotlin/app/nush/bot/commands/Nick.kt:
```diff
@@ -0,0 +1,96 @@
+                        ).toString()
+                        if (messageReaction.emoji.name == "❎") {
+                            clientStore.channels[clientStore.discord.createDM(
+                                CreateDM(
+                                    channelClient.getMessage(
+                                        messageReaction.messageId
+                                    ).usersMentioned[0].id
+                                )
```
This is common throughout both cases of the switch, you can use a variable to represent it instead

junron commented on src/main/kotlin/app/nush/bot/commands/Nick.kt:
```diff
@@ -0,0 +1,96 @@
+                channelClient.sendMessage("${guildClient.getMember(userId).nickname}(<@!$userId>) has requested to be renamed to \"$newName\". ✅ if you want to accept the rename request, ❎ if you do not want to accept the rename request")
+            renameMessage.react("✅")
+            renameMessage.react("❎")
+            reactionAdded { messageReaction ->
```
Move this to the `init` to prevent the handler from being added multiple times

junron commented on src/main/kotlin/app/nush/bot/commands/Nick.kt:
```diff
@@ -0,0 +1,96 @@
+                                channelClient.getMessage(messageReaction.messageId).content.indexOf("\"") + 1
+                            )
+                        ).toString()
+                        if (messageReaction.emoji.name == "❎") {
```
Use constants for emoji characters. Return before fetching message if emoji is neither cross or tick

junron commented on src/main/kotlin/app/nush/bot/commands/Verify.kt:
```diff
@@ -60,18 +61,101 @@ object Verify : Command {
-                config.memberRole
-            )
+            val channelClient = clientStore.channels[config.excoChannelId]
+            val isMember: (Member) -> Boolean = { it.email == user.email }
```
You can inline this:
```kotlin
members.any {
   it.email == user.email
}
```

junron commented on src/main/kotlin/app/nush/bot/commands/Verify.kt:
```diff
@@ -60,18 +61,101 @@ object Verify : Command {
     suspend fun userVerified(user: User, discordUserId: String) {
         val name = user.name.toLowerCase().capitalizeWords()
         with(bot) {
+            config.userIsAppVentureMember = false
```
Use a local variable instead

junron commented on src/main/kotlin/app/nush/bot/commands/Verify.kt:
```diff
@@ -60,18 +61,101 @@ object Verify : Command {
                 name
             )
-            clientStore.channels[channel.id].sendMessage("Welcome, $name")
+            reactionAdded { messageReaction ->
```
See comments on Nick.kt

junron commented on src/main/kotlin/app/nush/bot/commands/Nick.kt:
```diff
@@ -0,0 +1,96 @@
+                request(
+                    bot,
+                    authorId,
+                    words.drop(1).joinToString(separator = " ")
```
Ensure nickname is not empty.

zerver0 committed:
saved emotes in config

zerver0 committed:
add checks to make sure the message is sent by the bot and that it's an access requesting message and it's one of the 3 expected emotes
move event handler to init to prevent multiple additions

junron commented on src/main/kotlin/app/nush/bot/Config.kt:
```diff
@@ -11,7 +11,14 @@ data class Config(
+    val alumniRole: String,
+    val guestRole: String,
+    val excoChannelId: String,
+    val botId: String,
```
You can fetch and store this when the bot starts.

junron commented on src/main/kotlin/app/nush/bot/Config.kt:
```diff
@@ -11,7 +11,14 @@ data class Config(
+    val guestRole: String,
+    val excoChannelId: String,
+    val botId: String,
+    val hatEmote: String,
+    val cross: String,
+    val tick: String
```
You can store this as constants in your program

junron committed:
Remove botId from config

zerver0 committed:
add check to make sure there actually was a new requested name instead of nothing
move reaction handler to init to prevent multiple instances of the handler being added
weird bug of unable to use == to compare strings :/

zerver0 committed:
Merge branch 'role-allocation' of https://github.com/appventure-nush/appventure-bot into role-allocation

zerver0 committed:
move event handlers to init to prevent multiple additions of the handler
move expected emotes from config to individual event handlers
fix formatting

junron committed:
Tidy up code

junron commented on src/main/kotlin/app/nush/bot/commands/Nick.kt:
```diff
@@ -0,0 +1,123 @@
+                    }
+                    val guildClient = clientStore.guilds[config.guildId]
+                    val channelClient = clientStore.channels[config.excoChannelId]
+                    if (!guildClient.getMember(messageReaction.userId).user?.isBot!!) {
```
You just need to check that the `messageReaction.userId` != `botId`

junron replied:
You can flip the condition and return so the code is less indented

junron commented on src/main/kotlin/app/nush/bot/commands/Nick.kt:
```diff
@@ -0,0 +1,123 @@
+                                )
+                            ).toString()
+                        if (messageReaction.emoji.name == cross) {
+                            clientStore.channels[clientStore.discord.createDM(
```
The [`dmUser`](https://github.com/appventure-nush/appventure-bot/pull/7/files#diff-2c83c825a2f58f368981efef74d99186)  method has been added. You can replace all the `CreateDM`s with `dmUser`.

junron commented on src/main/kotlin/app/nush/bot/commands/Verify.kt:
```diff
@@ -24,7 +29,87 @@ object Verify : Command {
         with(prefix) {
             command("sendverify") {
-                sendVerifyMessage(bot, authorId)
+                userVerified(User("SU ZHENGCHONG", "h1810141@nushigh.edu.sg", 0), authorId)
```
Revert this to `sendVerifyMessage(bot, authorId)`

junron commented on src/main/kotlin/app/nush/bot/commands/Verify.kt:
```diff
@@ -24,7 +29,87 @@ object Verify : Command {
+                    }
+                    val guildClient = clientStore.guilds[config.guildId]
+                    val channelClient = clientStore.channels[config.excoChannelId]
+                    if (!guildClient.getMember(messageReaction.userId).user?.isBot!!) {
```
Just need to check that `userId` != `botId`

junron commented on src/main/kotlin/app/nush/bot/commands/Verify.kt:
```diff
@@ -24,7 +29,87 @@ object Verify : Command {
+                        if (accessRequestMessage.authorId != botId) return@reactionAdded
+                        when (messageReaction.emoji.name) {
+                            hatEmote -> {
+                                clientStore.channels[clientStore.discord.createDM(
```
Replace with `dmUser`

junron commented on src/main/kotlin/app/nush/bot/commands/Verify.kt:
```diff
@@ -24,7 +29,87 @@ object Verify : Command {
+        }
+        with(bot) {
+            reactionAdded { messageReaction ->
+                try {
```
Not required

zerver0 committed:
remove unnecessary try catch
replace all dm cases with dmUser
change check case on whether the user who reacted is the bot

zerver0 committed:
replace message type checks with in operator

junron committed:
Switch dmUser to extension function

junron committed:
Merge branch 'master' into role-allocation

junron committed:
Fix syntax error
